docs/source/blueprints/BP003.rst

```diff
@@ -0,0 +1,549 @@
+BP003
+#####
+
+:Number: BP003
+:Title: N(ext)Runner Task Phases
+:Author: Cleber Rosa <crosa@redhat.com>
+:Discussions-To: avocado-devel@redhat.com
+:Reviewers:
+:Created: 20-July-2020
+:Type: Architecture Blueprint
+:Status: Draft
+
+.. contents:: Table of Contents
+
+TL;DR
+*****
+
+The N(ext) Runner has been used as Avocado's runner for selftests for
+over a year.  The implementation used is based on the ``avocado nrun``
+command, that is, outside of the Avocado's traditional ``avocado run``
+entrypoint.  Under the hood, it means that the N(ext) Runner is not
+integrated well enough with an Avocado Job.
+
+A partial implementation of the N(ext) Runner integration with an
+Avocado Job is available at ``avocado/plugins/runner_nrunner.py`` but
+it has a number limitations.
+
+The N(ext) Runner executes tests as :class:`Tasks
+<avocado.core.nrunner.Task>`.  This blueprint describes the phases
+that a Task can be in throughout its life-cycle, and how the handling
+of these phases or states, will power the tests execution mechanism
+within the context of an Avocado Job.
+
+Motivations
+***********
+
+1. Explain how, with the distinguishing N(ext) Runner features, new
+   requirements are introduced with regards to its integration with an
+   Avocado Job.
+
+2. Propose an architecture based on the phases that an N(ext) Runner
+   Task can go through while running under an Avocado Job.
+
+3. Describe how the proposed architecture can power an implementation
+   suitable for the next Avocado LTS release (82.0), having feature
+   completeness when compared to the current runner, while still
+   making its distinguishing features available to users who opt in.
+
+4. Prove that the current runner can be removed without significant
+   user impact after the LTS release (within the 83.0 development
+   cycle), based on the feature completeness of the N(ext) Runner.
+
+5. Allow for future extension of the Task Phases architecture, such as
+   into a more capable and further reaching scheduler for Tasks.
+
+Requirements
+************
+
+Task Execution Requirements Verification
+----------------------------------------
+
+For a Task to actually be executed, there needs to be a minimal number
+of requirements present.  For instance, it's pointless to attempt to
+execute a Task of kind "custom" without either:
+
+1. An ``avocado-runner-custom`` runner script that is compatible with
+   the Avocado interface, OR
+
+2. A ``CustomRunner`` runner class that is compatible with the
+   :class:`avocado.core.nrunner.BaseRunner` interface
+
+Other types of Task Execution Requirements checks may be added in the
+future, but the core concept that a Task can not always be executed
+remains.
+
+Currently, as per the ``avocado nrun`` implementation, this
+verification is done in a synchronous way, and it's of limited
+visibility to the user.
+
+Requirements:
+
+1. The verification of one Task's requirement should not block other
+   Tasks from progressing to other phases.
```

Comment:
This phase is not dealing with the task execution itself, that is, it's not about a `TASK_RUNNING` phase, so I'm not sure if we should enforce/allow no-parallelism here.

Having said that, there may need to be some rate limiting mechanism to avoid the requirements check from exhausting resources. I mean, we should probably not run 100s of requirements verification "threads" at once.

Reply:
It does, thanks. But then I would like to move my question to the "TASK_RUNNING" section: Are we assuming that the tasks will start running in any case? Or only when parallel tasks is enabled? Are we assuming that all tasks are parallel?

Reply:
I think this is the big difference between the legacy runner and the N(ext) Runner. The legacy runner runs tests sequentially; the N(ext) Runner runs tests in parallel by default. If the user needs the tests to run sequentially, there is an option for that.

Reply:
Yes, I know @willianrampazzo. I'm just missing this in this document. Since that gave us the impression that everything will be in parallel.

Reply:
Okay, got it!